### PR TITLE
Minimize calls to ReadModules in UserSpaceInstrumentation

### PR DESCRIPTION
**Pull request by dpallotti**
`ReadModules` is an expensive operation: it reads and parses the maps, it opens
and parses object files, and it prints plenty of information to the logs (which
made this very visible).
Previously, we were calling it in every function that needed the modules. Also,
the number of calls was growing with the number of instrumented modules.
Now we try to call it as few times as possible and pass the result down. Of
course this adds a parameter to various functions.

Test: Capture with user space instrumentation and manual instrumentation. Verify
they work and verify from the logs that the number of calls to `ReadModules` is
now reduced.

### CODE REVIEW

**Comment by dpallotti** [comment]
@danielfenner Let me know what you think, I hope it's not controversial. I have been thinking of doing something like this for a while having worked on `ReadModules` a lot in the last months.

File: src/UserSpaceInstrumentation/FindFunctionAddress.cpp [hunk 1]
```diff
@@ -8,30 +8,25 @@
...
+  for (const orbit_grpc_protos::ModuleInfo& module : modules) {
     if (module.soname() == module_soname) {
       module_file_path = module.file_path();
       module_base_address = module.address_start();
     }
   }
   if (module_file_path.empty()) {
     return ErrorMessage(
-        absl::StrFormat("There is no module \"%s\" in process %d.", module_soname, pid));
+        absl::StrFormat("There is no module \"%s\" in the target process", module_soname));
```
Comment (dpallotti) [inline src/UserSpaceInstrumentation/FindFunctionAddress.cpp:29]: I removed the `pid` parameter from this function as now it was only used for this one message. As a result, I could also remove it from `FindFunctionAddressWithFallback`.

**Review by danielfenner — APPROVED** [review]
> Let me know what you think, I hope it's not controversial. I have been thinking of doing something like this for a while having worked on ReadModules a lot in the last months.

Thanks for cleaning this up! I wasn't aware of the cost of ReadModules. Especially dragging this out of the loop over the instrumented functions sgtm.

lgtm!

**Comment by dpallotti** [comment]
> the number of calls was growing with the number of instrumented functions.

Correction: this was growing with the number of modules with instrumented functions, because of `modules_from_path_`.

**Review by danielfenner — APPROVED** [review]
Handling of cache_of_modules_from_path looks better now, thanks! lgtm1

**Pull request merged**